src/plugins/spec/spec-new.c

```diff
@@ -0,0 +1,754 @@
+ *
+ * @return the extracted specification keys
+ */
+static KeySet * extractSpecKeys (KeySet * ks)
+{
+	KeySet * specKeys = ksNew (0, KS_END);
+	for (elektraCursor it = 0; it < ksGetSize (ks); it++)
+	{
+		Key * current = ksAtCursor (ks, it);
+		if (keyGetNamespace (current) == KEY_NS_SPEC)
+		{
+			ksAppendKey (specKeys, current);
+		}
+	}
+
+	Key * specKey = keyNew ("spec:/", KEY_END);
+	ksDel (ksCut (ks, specKey));
+	keyDel (specKey);
+
+	return specKeys;
+}
+
+/**
+ * Checks if the {@link specKey} is a array specification key.
+ *
+ * @param specKey the specification key to check if it is an array specification
+ * @return true  - if it is an array specification
+ * 	   false - if it is no array specification
+ */
+static bool isArraySpecification (Key * specKey)
+{
+	const char * keyWithoutNamespace = strchr (keyName (specKey), '/');
+
+	for (size_t i = 0; i < elektraStrLen (keyWithoutNamespace); i++)
+	{
+		if (keyWithoutNamespace[i] == '#')
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * Checks if this specification key contains an `_` in the array specification.
+ *
+ * @param specKey the specification key to check for underline if it is array specification
+ * @return true - if the array specification contains an `_`
+ * 	   false - if the array specification does not contain an `_`
+ */
+static bool containsUnderlineInArraySpec (Key * specKey)
+{
+	const char * keyWithoutNamespace = strchr (keyName (specKey), '/');
+	size_t len = elektraStrLen (keyWithoutNamespace);
+
+	for (size_t i = 0; i < len; i++)
+	{
+		if (keyWithoutNamespace[i] == '#' && (i != len && keyWithoutNamespace[i + 1] == '_'))
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * Checks if this specification key has a wildcard character in the key name.
+ *
+ * @param specKey the specification to check for if it has a wildcard character
+ * @return true - if the specification key has a wildcard character in the key name
+ * 	   false - if the specification does not contain a wildcard character
+ */
+static bool isWildcardSpecification (Key * specKey)
+{
+	const char * keyWithoutNamespace = strchr (keyName (specKey), '/');
+	for (size_t i = 0; i < elektraStrLen (keyWithoutNamespace); i++)
+	{
+		if (keyWithoutNamespace[i] == '_')
+		{
+			return true;
+		}
+	}
+
+	return false;
+}
+
+/**
+ * Creates the name of the array element.
+ *
+ * Example:
+ * 	arrayNumber = 2
+ * 	size = 2
+ *
+ * 	arrayElement = #2
+ *
+ * 	arrayNumber = 10
+ * 	size = 4
+ *
+ * 	arrayElement = #_10
+ *
+ * 	arrayNumber = 100
+ * 	size = 6
+ *
+ * 	arrayElement = #__100
+ *
+ * @param arrayElement the element to store the name in
+ * @param arrayNumber the array element number
+ * @param size the size of array
+ */
+static void createArrayElementName (char * arrayElement, int arrayNumber, int size)
+{
+	arrayElement[0] = '#';
+	for (int j = 1; j < (arrayNumber % 10); j++)
+	{
+		arrayElement[j] = '_';
+	}
+	sprintf (&arrayElement[size], "%d", arrayNumber);
```

Comment:
The only time you call this function, you pass the number of allocated bytes for `arrayElement` as the parameter `size`. So the maximum index would be `size-1`. And here you want to write something _past_ that?